File: 92-reverse-linked-list-ii/reverse-linked-list-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        ListNode *dummy_head = new ListNode(0, head);
        ListNode *prev = dummy_head, *curr = head;

        for(int i=0; i<left-1; i++){
            prev = prev->next;
            curr = curr->next;
        }

        for(int i=0; i<(right-left); i++){
            ListNode *temp = curr->next;
            curr->next = temp->next;
            temp->next = prev->next;
            prev->next = temp;
        }

        return dummy_head->next;
    }
};

File: 92-reverse-linked-list-ii/reverse-linked-list-ii.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseBetween(head *ListNode, left int, right int) *ListNode {
    if head.Next == nil || left == right {
        return head
    }

    prev, st, end := &ListNode{Val: 0, Next:head}, head, head

    for i := left; i > 1; i-- {
        prev = st
        st = st.Next
    }

    for i := right; i > 1; i-- {
        end = end.Next
        right--
    }

    // fmt.Println(prev, st, end)

    tmp := st.Next
    st.Next = end.Next
    prev.Next = end
    prev = st
    st = tmp

    for prev != end {
        tmp = st.Next
        st.Next = prev
        prev = st
        st = tmp
    }

    if left == 1 {
        return end
    }
    
    return head
}